src/dex/verified/pools/primary/primaryPoolMath.ts

```diff
@@ -0,0 +1,592 @@
+      sum += this.x1;
+    }
+
+    // All other a_n and x_n are stored as 20 digit fixed point numbers, so we convert the sum and a to this format.
+    sum *= 100n;
+    a *= 100n;
+
+    // Because further a_n are  20 digit fixed point numbers, we multiply by ONE_20 when dividing by them.
+
+    if (a >= this.a2) {
+      a = (a * BI_POWS[20]) / this.a2;
+      sum += this.x2;
+    }
+
+    if (a >= this.a3) {
+      a = (a * BI_POWS[20]) / this.a3;
+      sum += this.x3;
+    }
+
+    if (a >= this.a4) {
+      a = (a * BI_POWS[20]) / this.a4;
+      sum += this.x4;
+    }
+
+    if (a >= this.a5) {
+      a = (a * BI_POWS[20]) / this.a5;
+      sum += this.x5;
+    }
+
+    if (a >= this.a6) {
+      a = (a * BI_POWS[20]) / this.a6;
+      sum += this.x6;
+    }
+
+    if (a >= this.a7) {
+      a = (a * BI_POWS[20]) / this.a7;
+      sum += this.x7;
+    }
+
+    if (a >= this.a8) {
+      a = (a * BI_POWS[20]) / this.a8;
+      sum += this.x8;
+    }
+
+    if (a >= this.a9) {
+      a = (a * BI_POWS[20]) / this.a9;
+      sum += this.x9;
+    }
+
+    if (a >= this.a10) {
+      a = (a * BI_POWS[20]) / this.a10;
+      sum += this.x10;
+    }
+
+    if (a >= this.a11) {
+      a = (a * BI_POWS[20]) / this.a11;
+      sum += this.x11;
+    }
+
+    // a is now a small number (smaller than a_11, which roughly equals 1.06). This means we can use a Taylor series
+    // that converges rapidly for values of `a` close to one - the same one used in ln_36.
+    // Let z = (a - 1) / (a + 1).
+    // ln(a) = 2 * (z + z^3 / 3 + z^5 / 5 + z^7 / 7 + ... + z^(2 * n + 1) / (2 * n + 1))
+
+    // Recall that 20 digit fixed point division requires multiplying by ONE_20, and multiplication requires
+    // division by ONE_20.
+    const z = ((a - BI_POWS[20]) * BI_POWS[20]) / (a + BI_POWS[20]);
+    const z_squared = (z * z) / BI_POWS[20];
+
+    // num is the numerator of the series: the z^(2 * n + 1) term
+    let num = z;
+
+    // seriesSum holds the accumulated sum of each term in the series, starting with the initial z
+    let seriesSum = num;
+
+    // In each step, the numerator is multiplied by z^2
+    num = (num * z_squared) / BI_POWS[20];
+    seriesSum += num / 3n;
+
+    num = (num * z_squared) / BI_POWS[20];
+    seriesSum += num / 5n;
+
+    num = (num * z_squared) / BI_POWS[20];
+    seriesSum += num / 7n;
+
+    num = (num * z_squared) / BI_POWS[20];
+    seriesSum += num / 9n;
+
+    num = (num * z_squared) / BI_POWS[20];
+    seriesSum += num / 11n;
+
+    // 6 Taylor terms are sufficient for 36 decimal precision.
+
+    // Finally, we multiply by 2 (non fixed point) to compute ln(remainder)
+    seriesSum *= 2n;
+
+    // We now have the sum of all x_n present, and the Taylor approximation of the logarithm of the remainder (both
+    // with 20 decimals). All that remains is to sum these two, and then drop two digits to return a 18 decimal
+    // value.
+
+    return (sum + seriesSum) / 100n;
+  }
+}
```

Comment:
@Mohzcrea8me are we really using so many functions for primarypool math ?

Reply:
No for now, but that's the math from fixed.pool and normal math function I don't know how many we will need for verified.ts yet when I do it will be corrected